### ISSUE TITLE
Run tests with both deriving methods

### ISSUE THREAD

**Issue by treeowl** [linear-generics/linear-generics#4]
Currently, the tests always derive generics using Template Haskell. We should set them up to use either TH or `DerivingVia` with some CPP, and run them both ways. I don't really know the right way to do that. Do we need `TheTestTH.hs` and `TheTestDV.hs` that each set a macro and `#include "TheTest.hs"`, or is there some nicer way?

**Comment by treeowl** [comment]
@aspiwack, yes, via `GHC.Generics`, which turns out to be *really* unsafe for types with nonlinear fields. Fun times.

**Comment by aspiwack** [comment]
> which turns out to be _really_ unsafe for types with nonlinear fields. 

Good point.
